DOCS-2741: Iterate on Update and Manage Modules (#4035)
viamrobotics/docs: docs/operate/get-started/other-hardware/manage-modules.md

cost: "0"
---

After you [create and deploy a module](/operate/get-started/other-hardware/), you can update, delete, or change its visibility settings.

For information on pinning module deployments to versions, see [Module versioning](/operate/reference/module-configuration/#module-versioning).

## Update a module

Once your module is in the [Viam Registry](https://app.viam.com/registry), there are two ways to update it:

- [Update automatically](#update-automatically) using GitHub Actions: Recommended for ongoing projects with continuous integration (CI) workflows, or if you want to build for multiple platforms.

>   - If you enabled cloud build when you generated your module, the GitHub actions are already set up for you.

Author: Played with this locally and it looks odd because it ends up closer to the next bullet so going to leave as-is

- [Update manually](#update-manually) using the [Viam CLI](/dev/tools/cli/): Fine for small projects with one contributor.

### Update automatically

Use GitHub Actions to automatically build and deploy your new module version when you create a tag or release in GitHub:

> 1. Edit your module code and update the [`meta.json`](/operate/get-started/other-hardware/#metajson-reference) file if needed.

Author: ```suggestion
1. Edit your module code.
   If necessary, edit the [`meta.json`](/operate/get-started/other-hardware/#metajson-reference) file to reflect your changes.
   For example, if you added or changed the functionality of models, update the models object.
```
Like this?

> 1. Edit your module code and update the [`meta.json`](/operate/get-started/other-hardware/#metajson-reference) file if needed.

Author: Made it match the other one ✔️

> 1. Edit your module code and update the [`meta.json`](/operate/get-started/other-hardware/#metajson-reference) file if needed.

Reviewer: I mean that covers module funcitonality but I think people MUST update the meta.json the first time they set up automatic updates to set all attributes required for cloud build. Or has that changed?

> 1. Edit your module code and update the [`meta.json`](/operate/get-started/other-hardware/#metajson-reference) file if needed.

Author: You mean if they generated it with cloud build disabled?

> 1. Edit your module code and update the [`meta.json`](/operate/get-started/other-hardware/#metajson-reference) file if needed.

Reviewer: even with cloud build enabled you still must set at least one attribute to get cloud build to work. https://docs.viam.com/operate/get-started/other-hardware/#metajson-reference

> 1. Edit your module code and update the [`meta.json`](/operate/get-started/other-hardware/#metajson-reference) file if needed.

Author: Are you referring to the URL? If they followed the guide, they updated the URL in step 2 right above where you linked. I'll add it below as well though.

   {{% alert title="Note" color="note" %}}

   If you used `viam module generate` to create your module and enabled cloud build, all you need to do to trigger a new build is create a tag and publish a release in GitHub as you did when you [first published the module](/operate/get-started/other-hardware/#upload-your-module).

   {{% /alert %}}

1. If you did not use the Viam CLI generator and enable cloud build, you can set up one of the following GitHub actions up manually:

   {{< tabs >}}
   {{% tab name="build-action (Recommended)" %}}

   The `build-action` GitHub action provides a simple cross-platform build setup for multiple platforms: x86, ARM Linux, and macOS.

>    Add this to your GitHub workflow:

Reviewer: Unless people already know how GitHub Actiosn work, I think we need to link out to GH docs here on this. Or explain how to do this. I think we're also assuming here that they have already pushed to GitHub - that probably also needs clarifying

Author: Okay making some tweaks. I must say the GitHub UI is pretty intuitive--I hadn't gone through this workflow before recently and I found it easy to follow without too much documentation. I know I'm just a sample size of 1 but I want to keep this concise and we do have the link to more docs below, and i've added the link above too. LMK what you think of latest commit.

Reviewer: ok, could add it as a prerequisite but optional

Reviewer: For adding the workflow, they'd generally be creating a new file for the build action, I think. Either way this currently still requires you to know where workflows are and how to add them, easier to specify:
```suggestion
   In your repository, create a workflow file called <FILE>build.yml</FILE> in the <FILE>.github/workflows<FILE> directory:
```

cost: "0"
---

After you [create and upload a module](/operate/get-started/other-hardware/), you can update, delete, or change its visibility settings.

For information on pinning module deployments to versions, see [Module versioning](/operate/reference/module-configuration/#module-versioning).

## Update a module

Once your module is in the [Viam Registry](https://app.viam.com/registry), there are two ways to update it:

- [Update automatically](#update-automatically) using GitHub Actions: Recommended for ongoing projects with continuous integration (CI) workflows, or if you want to build for multiple platforms.

  - If you enabled cloud build when you generated your module, the GitHub Actions are already set up for you.

- [Update manually](#update-manually) using the [Viam CLI](/dev/tools/cli/): Fine for small projects with one contributor.

### Update automatically

Use [GitHub Actions](https://docs.github.com/actions) to automatically build and deploy your new module version when you create a tag or release in GitHub:

1. Edit your module code and update the [`meta.json`](/operate/get-started/other-hardware/#metajson-reference) file if needed.

>    For example, if you've changed the module's functionality, update the description in the `meta.json` file.

Author: ```suggestion
   For example, if you've changed the module's functionality, update the description in the `meta.json` file.

   Make sure the `url` field contains the URL to the GitHub repo that contains your module code.
```

> 1. Push your changes to the GitHub repository where you have your module code.

Author: ```suggestion
1. Push your changes to your module GitHub repository.
```

- [C++](https://github.com/viamrobotics/module-example-cpp)

  {{% /tab %}}
  {{< /tabs >}}

3. [Create an organization API key](/dev/tools/cli/#create-an-organization-api-key) with owner role:

   ```sh {class="command-line" data-prompt="$"}
   viam organizations api-key create --org-id <org-id> --name <key-name>
   ```

1. Add the key ID and value as GitHub repository secrets named `viam_key_id` and `viam_key_value`.

1. Push a tag or create a [release](https://docs.github.com/en/repositories/releasing-projects-on-github) in GitHub to trigger the build.
   Once the build is complete, the module will automatically update in the [Viam Registry](https://app.viam.com/registry), and the machines set to use the latest [version](/operate/reference/module-configuration/#module-versioning) of the module will automatically update to the new version.

### Update manually

Use the [Viam CLI](/dev/tools/cli/) to manually update your module:

1. Edit your module code and update the [`meta.json`](/operate/get-started/other-hardware/#metajson-reference) file if needed.
   For example, if you've changed the module's functionality, update the description in the `meta.json` file.

> 2. For Python modules only, package your files as an archive:

Reviewer: Since this may not apply to what people need to do, it may be worth at least signposting that that may not be what they need to run.
```suggestion
2. For Python modules only, package your files as an archive, for example:
```

   ```sh {class="command-line" data-prompt="$"}
   tar -czf module.tar.gz run.sh requirements.txt src
   ```

   Supply the path to the resulting archive file in the next step.

3. Upload to the Viam Registry:

   ```sh {class="command-line" data-prompt="$"}
   viam module upload --version <version> --platform <platform> <module-path>
   ```

   For example, `viam module upload --version 1.0.1 --platform darwin/arm64 my-module.tar.gz`.

When you `upload` a module, the command performs basic [validation](/dev/tools/cli/#upload-validation) of your module to check for common errors.

For more information, see the [`viam module` command](/dev/tools/cli/#module).

## Change module visibility

You can change the visibility of a module from public to private if:

> - you are an [owner](/manage/manage/rbac/) in the {{< glossary_tooltip term_id="organization" text="organization" >}} that owns the module, AND

Reviewer: ```suggestion
- you are an [owner](/manage/manage/rbac/) of the {{< glossary_tooltip term_id="organization" text="organization" >}} that owns the module, AND
```

1. Navigate to its page in the [**REGISTRY** section of the Viam app](https://app.viam.com/registry)
2. Click **Show previous versions** under the **Latest version** heading.

> 3. Hover next to the version you'd like to delete and click the trash icon.

Reviewer: ```suggestion
3. Hover on the architecture pill next to the version you'd like to delete and click the trash icon.
```
